### DISCUSSION TITLE
[BUG] selective activation recompute only decrease little of GPU memory usage during training

### DISCUSSION THREAD

**Discussion by bugm** [NVIDIA/Megatron-LM#1253] [Q&A] [answered]
**Describe the bug**
According to the paper https://arxiv.org/abs/2205.05198. The normal activation memory for a transformed based model in each layer can be calculated as 
![Image](https://github.com/user-attachments/assets/a9522623-43b3-4986-bf26-2ba807868527)
and with the selective activation recompute, it can be decreased to 
![Image](https://github.com/user-attachments/assets/5cfcbd37-990f-4c6e-9ede-f6c9294bf2f0)

with my training set, 
![Image](https://github.com/user-attachments/assets/15348b27-3c55-495a-88d0-1ddf65f60f04)
with tp =1 and pp =1, I expected when i use --recompute-activations, the GPU memory usage for storing activation should only be about 34 / (34+80)  = 30% of that with no activation recompute applied.

Here are some info about the GPU memory usage 
with --recompute-activations
![Image](https://github.com/user-attachments/assets/4c534f4b-064a-4b2d-8e50-66e0448c0505)

without   --recompute-activations
![Image](https://github.com/user-attachments/assets/fb31eb4e-c2e8-4d26-a926-2ef7500aaf18)

I notice the max_memory allocated during training only decreased from 25.52GB to 24.94 GB. 

**Expected behavior**
The max_memory allocated  during training should decrease more.

**Environment (please complete the following information):**
 - Megatron-LM commit ID
 - PyTorch 2.4.1 
 - CUDA version  12.5
 - NCCL version 2.20.5

**Additional context**
According to the formula above, with b = 12 s =1024 h =1024  L= 20 a=16 t=1, the original activation memory should be around 32GB,  plus the memory for model states  , which is about 7.3 GB for a  0.43B parameters model, which should be around 40GB even not take the temporary buffers and unusable fragment memory into account. That is much bigger than the max_memory allocated without activation recomputing, So I wonder the Megatron-LM has done some optimize here?
And why the max_memory allocated only changes little with/without --recompute-activations （use selective activation as default according to the doc）

**Comment by elliottnv** [comment]
Thanks for sharing the issue. We currently do not have bandwidth to investigate more, and we will rely on the community to explore more on the issue.

**Comment by wplf** [comment]
selective activation only recompute attention.
If you want to save more memory, you can try --full-recompute  and  recompute the whole transformer layer.

**Comment by jpy794** [comment, marked as answer]
`--recompute-activations` doesn't make any difference here because by default Megatron-LM uses TransformerEngine backend. And fused attention kernels like flash-attention always do selective activation checkpointing. You can set `--transformer-impl local`, and `--recompute-activations` would behave as described in paper. Maybe we should document this for clarity?
https://github.com/Dao-AILab/flash-attention/blob/641db759ab7168e472909bc9ff1eda4a329de34f/flash_attn/flash_attn_interface.py#L915

**Reply by bugm** [reply]
Thanks for your explanation!

**Reply by eliird** [reply]
I get that fused kernels are alreading doing the selective activation but changing the `recompute-granularity` to full also does not have any significant effect on memory usage.

**Reply by wplf** [reply]
Because most memory usage is released by flash attn.

**Reply by eliird** [reply]
![image](https://github.com/user-attachments/assets/57ae9954-b75f-44b1-ab75-800f77996679)
From the comment above and reading the flash attention paper again what I understood was flash attention should free memory equivalent to that of selective recompute. The image in the original paper shows that using full recomputations should reduce memory usage further by quite a big factor . 
![image](https://github.com/user-attachments/assets/3527f0e4-2b4c-42cb-960e-d4707930b708)
I am trying to run llama3-8b model with dp=1, tp=8 and pp=1 with batch size 1. And all I am seeing is change in memory from 21GB per GPU to 20.7GB. I am using 8xH100, which is quite different from what the paper reports. And I am trying to understand why?

**Reply by jpy794** [reply]
Do you also set `--recompute-method uniform` and  `--recompute-num-layers 1`?

**Reply by eliird** [reply]
Yes I do. I tried with `--recompute-num-layers 1 4 and 8`

**Reply by eliird** [reply]
I also tried disabling flash attention to test. Both  `selecetive` and `full` options have identical memory usage.

**Reply by jpy794** [reply]
Just tested with GPT3 6.7B，Megatron-LM core_r0.9.0. The result looks fine.
![selective-full](https://github.com/user-attachments/assets/5f74c73c-9cc5-4dc6-9d4c-cb00cb5dd606)

**Reply by eliird** [reply]
Thank you for the test. I was looking at the first 2 figures of the memory usage only. I was expecting a bit more of the drop in memory usage, but I realized with flash attention already in place there isn't much memory you can save anyway. 

Thank you once again for helping me understand.

**Comment by Listen-WLS** [comment]
i have a new question about GPU memory on training：https://github.com/NVIDIA/Megatron-LM/discussions/1486 
Have you ever had this problem before？

**Reply by eliird** [reply]
Yes, memory usage for me is the same as described in the issue. I always thought this might be because of the extra nccl kernels that need to be called to move data between GPUs. Usually pp communication is cheaper than tp. Because at some point you need to collect results on each GPU for normalization.

**Reply by eliird** [reply]
But take my word with a grain of salt, I never looked into it with much detail

**Reply by Listen-WLS** [reply]
thank u，my question is when the pp growing，mean-GPU memory  has decrease，i want to get a formulaic description for it

**Reply by eliird** [reply]
I dont think I am knowledgable enough to answer that question. Just some thoughts: it  willdepends on which layer the model is being pipelined. Assuming that you pipeline after end of each transformer block the communication would be similar so you could get a formula using some heuristic experiments. 
Look at usage when using TP=1 and PP=1. Then compare it to TP=1 and PP=2 annd then with PP=3 and so on.
